### DISCUSSION TITLE
Is there a way to draw a rectangle with a shader directly to a viewport?

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#7541] [2D]
I'd like to design a painting program and am struggling trying to figure out if Godot will let me draw shapes directly to surfaces or if my only choice is to create a bunch of nodes and wait for the next draw call for the Godot engine to do it.  

Basically I want to create some sort of texture or viewport that will hold the data for a layer of my image.  I then want to draw several rectangles representing the points of a pen stroke onto this buffer.  I want to be able to have these rectangles drawn using a shader so I can have control over properties such as the brush head's texture, rotation and opacity and how it is composited into the back buffer.

I could do this by creating a set of sprites, adding them to a viewport and then waiting for the viewport to update, but if there is a way to draw them more directly I think that would be more efficient.  I did come across this thread which seems to suggest there is a way, but it does not describe how exactly they are drawing the sprite to their surface.  https://godotforums.org/d/22141-blitting-textures/10

**Comment by Calinou** [comment]
Thanks for the proposal! Consolidating in https://github.com/godotengine/godot-proposals/issues/7379.
